### PR TITLE
Improve registers handling in ESIL dfg ##esil

### PR DESCRIPTION
**Pull request by condret**
It already works, but I'd like to do a bit more cleanup before merge.

### CODE REVIEW

File: libr/anal/esil_dfg.c [hunk 1]
```diff
@@ -26,13 +31,380 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+	if (!dfg || !sdb_num_exists (dfg->regs, reg)) {
+		return false;
+	}
+	EsilDFGRegVar *rv = R_NEW0 (EsilDFGRegVar);
+	if (!rv) {
+		return false;
+	}
+	const ut64 v = sdb_num_get (dfg->regs, reg, NULL);
+	rv->from = (v & 0xffffffff00000000) >> 32;
+	rv->to = v & 0xffffffff;
```
Comment (radare) [inline libr/anal/esil_dfg.c]: use UT32_MAX;

File: libr/anal/esil_dfg.c [hunk 2]
```diff
@@ -26,13 +31,380 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+static bool _edf_reg_set (RAnalEsilDFG *dfg, const char *reg, RGraphNode *node) {
+	if (!dfg || !sdb_num_exists (dfg->regs, reg)) {
+		return false;
+	}
+	EsilDFGRegVar *rv = R_NEW0 (EsilDFGRegVar);
+	if (!rv) {
+		return false;
+	}
+	const ut64 v = sdb_num_get (dfg->regs, reg, NULL);
+	rv->from = (v & 0xffffffff00000000) >> 32;
```
Comment (radare) [inline libr/anal/esil_dfg.c]: use UT32_MAX << 32 or maybe define UT32H_MAX or so

File: libr/anal/esil_dfg.c [hunk 3]
```diff
@@ -26,13 +31,380 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+
+// first handle the simple cases without intersection
+	if (rv_incoming->to < rv_in->from) {
+		return -1;
+	}
+	if (rv_in->to < rv_incoming->from) {
+		return 1;
+	}
+
+#if 0
```
Comment (radare) [inline libr/anal/esil_dfg.c]: use // for comment. if0 is for code

File: libr/anal/esil_dfg.c [hunk 4]
```diff
@@ -26,14 +30,373 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+
+// first handle the simple cases without intersection
+	if (rv_incoming->to < rv_in->from) {
+		return -1;
+	}
+	if (rv_in->to < rv_incoming->from) {
+		return 1;
+	}
+
+/*
```
Comment (radare) [inline libr/anal/esil_dfg.c]: remove this comment

File: libr/anal/esil_dfg.c [hunk 5]
```diff
@@ -26,14 +30,373 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+	r_rbtree_cont_insert (dfg->reg_vars, rv, _rv_ins_cmp, NULL);
+	return true;
 }
 
+
+static int _rv_find_cmp (void *incoming, void *in, void *user) {
+	EsilDFGRegVar *rv_incoming = (EsilDFGRegVar *)incoming;
+	EsilDFGRegVar *rv_in = (EsilDFGRegVar *)in;
+
+// first handle the simple cases without intersection
```
Comment (radare) [inline libr/anal/esil_dfg.c]: align, but i would remove the comment too

File: libr/anal/esil_dfg.c [hunk 6]
```diff
@@ -26,14 +30,373 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+	// rbtree api does sadly not allow deleting multiple items at once :(
+		rv = r_queue_dequeue (dfg->todo);
+		r_rbtree_cont_delete (dfg->reg_vars, rv, _rv_del_alloc_cmp, dfg);
+		if (dfg->insert) {
+			r_rbtree_cont_insert (dfg->reg_vars, dfg->insert, _rv_ins_cmp, NULL);
+			dfg->insert = NULL;
+		}
+		free (rv);
+	}
+	rv = R_NEW0 (EsilDFGRegVar);
```
Comment (radare) [inline libr/anal/esil_dfg.c]: nullchk

File: libr/anal/esil_dfg.c [hunk 7]
```diff
@@ -26,14 +30,373 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
-static void* _edf_reg_get(RAnalEsilDFG *edf, const char *reg) {
-	return edf ? sdb_ptr_get (edf->latest_nodes, reg, 0) : NULL;
+static int _rv_ins_cmp (void *incoming, void *in, void *user) {
+	EsilDFGRegVar *rv_incoming = (EsilDFGRegVar *)incoming;
+	EsilDFGRegVar *rv_in = (EsilDFGRegVar *)in;
+	return rv_incoming->from - rv_in->from;
+}
+
+static bool _edf_reg_set (RAnalEsilDFG *dfg, const char *reg, RGraphNode *node) {
+	if (!dfg || !sdb_num_exists (dfg->regs, reg)) {
```
Comment (radare) [inline libr/anal/esil_dfg.c]: r_return_val_if_fail (dfg && reg && node, false)

File: libr/anal/esil_dfg.c [hunk 8]
```diff
@@ -26,14 +30,373 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+		rv->from = rv_in->to + 1;
+		r_queue_enqueue (todo, rv);
+		return 0;
+	}
+
+/*
+   =incoming=
+  =in=
+
+similar to the previous case, but this time only enqueue 1 half
```
Comment (radare) [inline libr/anal/esil_dfg.c]: just one line comment with //

File: libr/anal/esil_dfg.c [hunk 9]
```diff
@@ -26,14 +30,373 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+*/
+	if (rv_in->from <= rv_incoming->from && rv_in->to < rv_incoming->to) {
+		EsilDFGRegVar *rv = R_NEW (EsilDFGRegVar);
+		rv[0] = rv_incoming[0];
+		rv->from = rv_in->to + 1;
+		r_queue_enqueue (todo, rv);
+		return 0;
+	}
+
+/*
```
Comment (radare) [inline libr/anal/esil_dfg.c]: remove comment

File: libr/anal/esil_dfg.c [hunk 10]
```diff
@@ -26,14 +30,373 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+     =incoming=
+=========in=========
+*/
+	if (rv_in->from <= rv_incoming->from && rv_incoming->to <= rv_in->to) {
+		return 0;
+	}
+
+/*
```
Comment (radare) [inline libr/anal/esil_dfg.c]: one line commment

File: libr/anal/esil_dfg.c [hunk 11]
```diff
@@ -26,14 +30,373 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+*/
+	RQueue *todo = (RQueue *)user;
+	if (rv_incoming->from < rv_in->from && rv_in->to < rv_incoming->to) {
+		// lower part
+		EsilDFGRegVar *rv = R_NEW (EsilDFGRegVar);
+		rv[0] = rv_incoming[0];
+		rv->to = rv_in->from - 1;
+		r_queue_enqueue (todo, rv);
+		// upper part
+		rv = R_NEW (EsilDFGRegVar);
```
Comment (radare) [inline libr/anal/esil_dfg.c]: nullchk

File: libr/anal/esil_dfg.c [hunk 12]
```diff
@@ -26,14 +30,373 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+	}
+
+/*
+   =incoming=
+  =in=
+
+similar to the previous case, but this time only enqueue 1 half
+*/
+	if (rv_in->from <= rv_incoming->from && rv_in->to < rv_incoming->to) {
+		EsilDFGRegVar *rv = R_NEW (EsilDFGRegVar);
```
Comment (radare) [inline libr/anal/esil_dfg.c]: nullchk and maybe r_new0 ? will be better to have esil_dfg_regvar_new(rv_incoming[0], rv_in->to + 1); (and the _free())

File: libr/anal/esil_dfg.c [hunk 13]
```diff
@@ -26,14 +30,373 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+*/
+	EsilDFGRegVar *rv = R_NEW (EsilDFGRegVar);
+	rv[0] = rv_incoming[0];
+	rv->to = rv_in->from - 1;
+	r_queue_enqueue (todo, rv);
+	return 0;
+}
+
+static RGraphNode *_edf_origin_reg_get(RAnalEsilDFG *dfg, const char *reg) {
+	if (!dfg || !sdb_num_exists (dfg->regs, reg)) {
```
Comment (radare) [inline libr/anal/esil_dfg.c]: add `r_return_val_if_fail (dfg && reg, false)`

File: libr/anal/esil_dfg.c [hunk 14]
```diff
@@ -26,14 +30,373 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+	RAnalEsilDFGNode *_origin_reg_node = r_anal_esil_dfg_node_new (dfg, reg);
+	r_strbuf_appendf (_origin_reg_node->content, ":var_%d", dfg->idx++);
+	_origin_reg_node->type = R_ANAL_ESIL_DFG_BLOCK_VAR;
+	origin_reg_node = r_graph_add_node (dfg->flow, _origin_reg_node);
+	r_graph_add_edge (dfg->flow, reg_node, origin_reg_node);
+	sdb_ptr_set (dfg->regs, origin_reg, origin_reg_node, 0);
+	free (origin_reg);
+	return origin_reg_node;
+}
+
```
Comment (radare) [inline libr/anal/esil_dfg.c]: empty line

File: libr/anal/esil_dfg.c [hunk 15]
```diff
@@ -26,14 +30,373 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+
+static RGraphNode *_edf_reg_get(RAnalEsilDFG *dfg, const char *reg) {
+	if (!dfg || !sdb_num_exists (dfg->regs, reg)) {
```
Comment (radare) [inline libr/anal/esil_dfg.c]: add r_return_val_if_fail (dfg && reg && node, false)

File: libr/anal/esil_dfg.c [hunk 16]
```diff
@@ -26,14 +30,373 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+	}
+	EsilDFGRegVar *rv = R_NEW0 (EsilDFGRegVar);
+	if (!rv) {
+		return NULL;
+	}
+	const ut64 v = sdb_num_get (dfg->regs, reg, NULL);
+	rv->from = (v & (UT64_MAX ^ UT32_MAX)) >> 32;
+	rv->to = v & UT32_MAX;
+	r_queue_enqueue (dfg->todo, rv);
+	RQueue *parts = r_queue_new (8);
```
Comment (radare) [inline libr/anal/esil_dfg.c]: null chk

File: libr/anal/esil_dfg.c [hunk 17]
```diff
@@ -26,14 +30,373 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+			part_rv = R_NEW (EsilDFGRegVar);
+			part_rv[0] = rv[0];
+			part_rv->node = reg_node;
+			r_rbtree_cont_insert (dfg->reg_vars, part_rv, _rv_ins_cmp, NULL);
+			//enqueue for later merge
+			r_queue_enqueue (parts, reg_node);
+		} else {
+			//initial regnode was already created
+			//only need to insert in the tree
+			part_rv = R_NEW (EsilDFGRegVar);
```
Comment (radare) [inline libr/anal/esil_dfg.c]: nullchk

File: libr/anal/esil_dfg.c [hunk 18]
```diff
@@ -26,14 +30,373 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+	reg_node = NULL;	// is this needed?
+	switch (parts->size) {
+	case 0:
+		break;
+	case 1:
+		reg_node = r_queue_dequeue (parts);
+		break;
+	default:
+		{
+			RAnalEsilDFGNode *_reg_node = r_anal_esil_dfg_node_new (dfg, "merge to ");
```
Comment (radare) [inline libr/anal/esil_dfg.c]: nullchk

File: libr/anal/esil_dfg.c [hunk 19]
```diff
@@ -26,14 +30,373 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+		}
+		do {
+			r_graph_add_edge (dfg->flow, r_queue_dequeue(parts), reg_node);
+		} while (!r_queue_is_empty (parts));
+		break;
+	}
+	r_queue_free (parts);
+	return reg_node;
+}
+
```
Comment (radare) [inline libr/anal/esil_dfg.c]: emptmy line

File: libr/anal/esil_dfg.c [hunk 20]
```diff
@@ -492,7 +831,10 @@ static bool edf_consume_1_use_old_new_push_1(RAnalEsil *esil, const char *op_str
 	return r_anal_esil_push (esil, r_strbuf_get (result->content));
 }
 
-R_API RAnalEsilDFG *r_anal_esil_dfg_new() {
+R_API RAnalEsilDFG *r_anal_esil_dfg_new(RReg *regs) {
+	if (!regs) {
```
Comment (radare) [inline libr/anal/esil_dfg.c]: r_return_val_if_fail (regs, NULL)

File: libr/anal/esil_dfg.c [hunk 21]
```diff
@@ -502,12 +844,40 @@ R_API RAnalEsilDFG *r_anal_esil_dfg_new() {
...
+	dfg->reg_vars = r_rbtree_cont_newf (free);
+	if (!dfg->reg_vars) {
+		r_queue_free (dfg->todo);
+		sdb_free (dfg->regs);
 		r_graph_free (dfg->flow);
 		free (dfg);
 		return NULL;
 	}
+
+// this is not exactly necessary
```
Comment (radare) [inline libr/anal/esil_dfg.c]: indent comments

File: libr/anal/esil_dfg.c [hunk 22]
```diff
@@ -26,12 +30,439 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+split in into 2 and reinsert the second half (in2)
+shrink first half (in1)
+
+     =incoming=
+=in1=          =in2=
+
+*/
+
+	if (rv_in->from < rv_incoming->from && rv_incoming->to < rv_in->to) {
+#if DFG_DEBUG
```
Comment (radare) [inline libr/anal/esil_dfg.c]: use anal->verbose or define D as if(0) or if(1) depending on DFG_DEBUG if u want, but personally i prefer to use frida to have such debug emssages

File: libr/anal/esil_dfg.c [hunk 23]
```diff
@@ -26,12 +30,439 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+		eprintf ("del_alloc_cmp: case 3 - return 1\n");
+#endif
+		EsilDFGRegVar *rv = R_NEW (EsilDFGRegVar);
```
Comment (radare) [inline libr/anal/esil_dfg.c]: missing nullchk

File: libr/anal/esil_dfg.c [hunk 24]
```diff
@@ -26,12 +30,439 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+		r_queue_enqueue (dfg->todo, rv);
+		// upper part
+		rv = R_NEW (EsilDFGRegVar);
+		rv[0] = rv_incoming[0];
+		rv->from = rv_in->to + 1;
+		r_queue_enqueue (dfg->todo, rv);
+		return 0;
+	}
+
+/*
```
Comment (radare) [inline libr/anal/esil_dfg.c]: kill those ugly commments

File: libr/anal/esil_dfg.c [hunk 25]
```diff
@@ -26,12 +30,439 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+#if DFG_DEBUG
+		eprintf ("del_alloc_cmp: case 4 - return 0\n");
+#endif
+		// lower part
+		EsilDFGRegVar *rv = R_NEW (EsilDFGRegVar);
+		rv[0] = rv_incoming[0];
+		rv->to = rv_in->from - 1;
+		r_queue_enqueue (dfg->todo, rv);
+		// upper part
+		rv = R_NEW (EsilDFGRegVar);
```
Comment (radare) [inline libr/anal/esil_dfg.c]: nullchk

File: libr/anal/esil_dfg.c [hunk 26]
```diff
@@ -26,12 +30,439 @@ void _dfg_node_free (RAnalEsilDFGNode *free_me) {
...
+		eprintf ("del_alloc_cmp: return 0 to delete [%d-%d]\n", rv_in->from, rv_in->to);
+#endif
+		r_queue_enqueue (dfg->todo, rv);
+#if DFG_DEBUG
+		eprintf ("del_alloc_cmp: enqueue [%d-%d] for deletion\n", rv->from, rv->to);
+#endif
+		return 0;
+	}
+
+/*
```
Comment (radare) [inline libr/anal/esil_dfg.c]: momre shitty comments

File: libr/anal/esil_dfg.c (hunk 24 above)
Comment (condret) [reply libr/anal/esil_dfg.c]: why? they explain what the callback is doing, and how it is doing it.

**Comment by radare** [comment]
please remove all those shitty comments asap

File: libr/anal/esil_dfg.c (hunk 16 above)
Comment (condret) [reply libr/anal/esil_dfg.c]: ok

File: libr/anal/esil_dfg.c (hunk 8 above)
Comment (condret) [reply libr/anal/esil_dfg.c]: why? why start every line with // ? that would be almost as bad as python

File: libr/anal/esil_dfg.c (hunk 11 above)
Comment (condret) [reply libr/anal/esil_dfg.c]: cannot do null checks in this callback rbtree does not allow aborting operation

File: libr/anal/esil_dfg.c (hunk 6 above)
Comment (condret) [reply libr/anal/esil_dfg.c]: can't

File: libr/anal/esil_dfg.c (hunk 18 above)
Comment (condret) [reply libr/anal/esil_dfg.c]: cant

File: libr/anal/esil_dfg.c (hunk 12 above)
Comment (condret) [reply libr/anal/esil_dfg.c]: really no need for R_NEW0 here, the todo-queue is only used to store further searchs here the rv->node is never derefed

**Pull request merged**